### PR TITLE
Implement Progenesis and show amount recouped

### PR DESCRIPTION
**Pull request by Regisle**
Implements Progenesis which applies before petrified blood and does not stack addatively, these are done using the EHP calc helper function to figure out roughly how much damage you take, these should be used as ratios compared to each other or compared to enemy damage config and not the actual values that will always be present

This also fixes a few issues with petrified blood, there is another issue that is not fixed but am unsure on the cause for now and will make an issue after this is merged

image

The new subsection is minimised by default but shows net recoup in header

### CODE REVIEW

**Comment by QuickStick123** [comment]
fixes https://github.com/PathOfBuildingCommunity/PathOfBuilding/issues/5248

**Comment by deathbeam** [comment]
Is the Petrified Blood issue you are mentioning here this? 
image

As that EHP jump is quite big even thoug I dont think its actually the case. https://pobb.in/u/thedeathbeam/pf-poison-scourge-arrow pob for reference just slap it there

**Comment by Regisle** [comment]
> Is the Petrified Blood issue you are mentioning here this? 

uh no, thats a new Issue, guess Ill try and debug that later

File: src/Modules/CalcDefence.lua [hunk 1]
```diff
@@ -1707,6 +1728,16 @@ function calcs.defence(env, actor)
...
 					end
+					-- soul link is not implemented for now
+					if soulLink > 0 then
+						local tempDamage = m_min(Damage[damageType] * output["SoulLinkMitigation"] / 100, soulLink)
+						soulLink = soulLink - tempDamage
+						Damage[damageType] = Damage[damageType] - tempDamage
+					end
+					-- frost shield and soul link does not count as you taking damage
+					if DamageIn["TrackLifeLoss"] then
+						output[damageType.."LifeLossLost"] = output[damageType.."LifeLossLost"] + Damage[damageType]
```
Comment (QuickStick123) [inline src/Modules/CalcDefence.lua:1739]:
```suggestion
					if DamageIn["TrackDamageTaken"] then
						output[damageType.."DamageTaken"] = output[damageType.."DamageTaken"] + Damage[damageType]
```
This has nothing to do with life using this naming is confusing.

File: src/Modules/CalcDefence.lua [hunk 2]
```diff
@@ -2034,24 +2083,73 @@ function calcs.defence(env, actor)
 		end
 	end
 	
+	-- recoup
+	if output["anyRecoup"] > 0 then
```
Comment (QuickStick123) [inline src/Modules/CalcDefence.lua:2087]: This should be re writeable with a for loop which should generalise some of this and be a bit neater.

File: src/Modules/CalcSections.lua [hunk 3]
```diff
@@ -2015,15 +2015,75 @@ return {
...
+			{ label = "Recovery modifiers", modName = "ManaRecoveryRate" },
+			{ label = "FasterRecoup", modName = "3SecondRecoup" },
+		},
+		{ format = "{0:output:EnergyShieldRecoupRecoveryAvg}",
+			{ breakdown = "EnergyShieldRecoupRecoveryAvg" },
+			{ label = "Sources", modName = { "EnergyShieldRecoup", "ElementalEnergyShieldRecoup" } },
+			{ label = "Recovery modifiers", modName = "EnergyShieldRecoveryRate" },
+			{ label = "FasterRecoup", modName = "3SecondRecoup" },
+		},
+	},
```
Comment (QuickStick123) [inline src/Modules/CalcSections.lua:2066]:
```suggestion
	{ label = "Recoup Max.", haveOutput = "anyRecoup", 
		{ format = "{0:output:LifeRecoupRecoveryMax}",
			{ breakdown = "LifeRecoupRecoveryMax" },
			{ label = "Sources", modName = { "LifeRecoup", "PhysicalLifeRecoup", "LightningLifeRecoup", "ColdLifeRecoup", "FireLifeRecoup", "ChaosLifeRecoup" } },
			{ label = "Recovery modifiers", modName = "LifeRecoveryRate" },
			{ label = "Faster Recoup", modName = "3SecondRecoup" },
		},
		{ format = "{0:output:ManaRecoupRecoveryMax}",
			{ breakdown = "ManaRecoupRecoveryMax" },
			{ label = "Sources", modName = "ManaRecoup" },
			{ label = "Recovery modifiers", modName = "ManaRecoveryRate" },
			{ label = "Faster Recoup", modName = "3SecondRecoup" },
		},
		{ format = "{0:output:EnergyShieldRecoupRecoveryMax}",
			{ breakdown = "EnergyShieldRecoupRecoveryMax" },
			{ label = "Sources", modName = { "EnergyShieldRecoup", "ElementalEnergyShieldRecoup" } },
			{ label = "Recovery modifiers", modName = "EnergyShieldRecoveryRate" },
			{ label = "Faster Recoup", modName = "3SecondRecoup" },
		},
	},
	{ label = "Recoup Avg.", haveOutput = "anyRecoup", 
		{ format = "{0:output:LifeRecoupRecoveryAvg}",
			{ breakdown = "LifeRecoupRecoveryAvg" },
			{ label = "Sources", modName = { "LifeRecoup", "PhysicalLifeRecoup", "LightningLifeRecoup", "ColdLifeRecoup", "FireLifeRecoup", "ChaosLifeRecoup" } },
			{ label = "Recovery modifiers", modName = "LifeRecoveryRate" },
			{ label = "Faster Recoup", modName = "3SecondRecoup" },
		},
		{ format = "{0:output:ManaRecoupRecoveryAvg}",
			{ breakdown = "ManaRecoupRecoveryAvg" },
			{ label = "Sources", modName = "ManaRecoup" },
			{ label = "Recovery modifiers", modName = "ManaRecoveryRate" },
			{ label = "Faster Recoup", modName = "3SecondRecoup" },
		},
		{ format = "{0:output:EnergyShieldRecoupRecoveryAvg}",
			{ breakdown = "EnergyShieldRecoupRecoveryAvg" },
			{ label = "Sources", modName = { "EnergyShieldRecoup", "ElementalEnergyShieldRecoup" } },
			{ label = "Recovery modifiers", modName = "EnergyShieldRecoveryRate" },
			{ label = "Faster Recoup", modName = "3SecondRecoup" },
		},
	},
```

File: src/Modules/CalcDefence.lua [hunk 4]
```diff
@@ -1397,28 +1401,37 @@ function calcs.defence(env, actor)
...
 				s_format("Total life protected:"),
 			}
+			if output["preventedLifeLoss"] ~= 0 then
+				t_insert(breakdown["preventedLifeLossTotal"], s_format("%.2f ^8(portion taken over 4 seconds instead)", output["preventedLifeLoss"] / 100))
+			end
 			if portionLife ~= 1 then
-				t_insert(breakdown["preventedLifeLoss"], s_format("%.2f ^8(initial portion taken from petrified blood)", output["preventedLifeLoss"] / portionLife / 100))
-				t_insert(breakdown["preventedLifeLoss"], s_format("* %.2f ^8(portion of life on low life)", portionLife))
-				t_insert(breakdown["preventedLifeLoss"], s_format("= %.2f ^8(final portion taken from petrified blood)", output["preventedLifeLoss"] / 100))
-				t_insert(breakdown["preventedLifeLoss"], s_format(""))
```
Comment (QuickStick123) [inline src/Modules/CalcDefence.lua:1417]: I feel communication of where the 0.16 number comes from would be useful to display in the breakdown also show these two parts being added would also be nice along side how you subtract from one for the end number.
Your calculations all look correct to me though.
image

**Review by LocalIdentity — APPROVED** [review]

**Pull request merged**